File: frontend/src/engine/capabilities/unpivot.ts
/**
 * unpivot – Convert a wide table into a tall (tidy) format.
 *
 * Example:
 *   Name | Jan | Feb | Mar          Name | Month | Value
 *   Alice|  10 |  20 |  30    →     Alice| Jan   | 10
 *   Bob  |   5 |  15 |  25          Alice| Feb   | 20
 *                                   ...
 *
 * The first `idColumns` columns are kept as-is on every output row.
 * The remaining columns become rows with two new columns: variable + value.
 */

import { CapabilityMeta, UnpivotParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "unpivot",
  description: "Unpivot (melt) a wide table to tall/tidy format",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: UnpivotParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const {
    sourceRange,
    idColumns,
    outputRange,
    variableColumnName = "Attribute",
    valueColumnName = "Value",
  } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would unpivot ${sourceRange} (${idColumns} id col(s)) to ${outputRange}` };
  }

  options.onProgress?.("Reading source table...");
  const srcRng = resolveRange(context, sourceRange);
  srcRng.load("values");
  await context.sync();

  const data = (srcRng.values ?? []) as (string | number | boolean | null)[][];
  if (data.length < 2) return { stepId: "", status: "success", message: "Not enough rows to unpivot." };

  const headers = data[0];
  const idHeaders = headers.slice(0, idColumns);
  const valueHeaders = headers.slice(idColumns);

  options.onProgress?.(`Unpivoting ${data.length - 1} rows × ${valueHeaders.length} value columns...`);

  // Build output rows: [id1, id2, ..., variableCol, valueCol]
  const outHeaders = [...idHeaders.map(String), variableColumnName, valueColumnName];
  const outRows: (string | number | boolean | null)[][] = [outHeaders];

  for (let r = 1; r < data.length; r++) {
    const idVals = data[r].slice(0, idColumns);
    for (let c = 0; c < valueHeaders.length; c++) {
      outRows.push([...idVals, String(valueHeaders[c]), data[r][idColumns + c]]);
    }
  }

  const outRng = resolveRange(context, outputRange);
  outRng.getResizedRange(outRows.length - 1, outHeaders.length - 1).values = outRows as any;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Unpivoted ${data.length - 1} rows into ${outRows.length - 1} rows (${valueHeaders.length} value columns → rows)`,
    outputs: { outputRange },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/unpivot.py
"""unpivot — convert a wide table into a long (tidy) format.

The first ``idColumns`` columns are kept as-is on every output row. Remaining
columns are unpivoted into two new columns (variable, value), so one input row
becomes (cols - idColumns) output rows.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    source = params.get("sourceRange")
    output = params.get("outputRange")
    id_columns = params.get("idColumns")
    variable_col_name = params.get("variableColumnName") or "Attribute"
    value_col_name = params.get("valueColumnName") or "Value"

    if not source or not output:
        return {
            "status": "error",
            "message": "unpivot requires 'sourceRange' and 'outputRange'.",
        }
    if id_columns is None:
        return {"status": "error", "message": "unpivot requires 'idColumns'."}

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": (
                f"Would unpivot {source} ({id_columns} id col(s)) to {output}."
            ),
        }

    try:
        src_rng = resolve_range(ctx.workbook_handle, source)
        raw = src_rng.value
        if raw is None:
            return {"status": "success", "message": "Not enough rows to unpivot.", "outputs": {}}
        if not isinstance(raw, list):
            data: list[list[Any]] = [[raw]]
        elif raw and not isinstance(raw[0], list):
            rows_, _ = src_rng.shape
            data = [list(raw)] if rows_ == 1 else [[v] for v in raw]
        else:
            data = [list(r) for r in raw]

        if len(data) < 2:
            return {"status": "success", "message": "Not enough rows to unpivot.", "outputs": {}}

        id_count = int(id_columns)
        headers = data[0]
        id_headers = [str(h) if h is not None else "" for h in headers[:id_count]]
        value_headers = headers[id_count:]

        out_headers: list[Any] = [*id_headers, variable_col_name, value_col_name]
        out_rows: list[list[Any]] = [out_headers]

        for r in range(1, len(data)):
            row = data[r]
            id_vals = list(row[:id_count])
            # Pad id_vals if row is short.
            if len(id_vals) < id_count:
                id_vals.extend([None] * (id_count - len(id_vals)))
            for c in range(len(value_headers)):
                source_col = id_count + c
                value = row[source_col] if source_col < len(row) else None
                out_rows.append([*id_vals, str(value_headers[c]) if value_headers[c] is not None else "", value])

        rows = len(out_rows)
        cols = len(out_headers)
        out_rng = resolve_range(ctx.workbook_handle, output)
        out_rng.resize(rows, cols).value = out_rows
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"unpivot failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": (
            f"Unpivoted {len(data) - 1} rows into {rows - 1} rows "
            f"({len(value_headers)} value columns → rows)."
        ),
        "outputs": {"outputRange": output},
    }


registry.register("unpivot", handler, mutates=True)
